File: services/indexer/repo_allowlist.py
"""Repository URL allowlist for indexer (SEC-03)."""

from __future__ import annotations

import logging
import os
import re
from urllib.parse import urlparse

logger = logging.getLogger(__name__)


def _parse_csv(name: str) -> list[str]:
    raw = (os.environ.get(name) or "").strip()
    if not raw:
        return []
    return [p.strip() for p in raw.split(",") if p.strip()]


def _parse_path_prefixes() -> list[str]:
    raw = (os.environ.get("INDEXER_ALLOWED_FILE_PATH_PREFIXES") or "/fixtures").strip()
    parts = [p.strip() for p in raw.split(",") if p.strip()]
    return parts or ["/fixtures"]


def normalize_github_https(repo_url: str) -> str | None:
    """Return canonical https://github.com/owner/repo (lowercase, no .git) or None."""
    u = repo_url.strip()
    if not u:
        return None
    u = re.sub(r"/+$", "", u)
    u = re.sub(r"\.git$", "", u, flags=re.IGNORECASE)
    m = re.match(r"^git@github\.com:(.+)$", u, flags=re.IGNORECASE)
    if m:
        path = m.group(1).strip()
        u = f"https://github.com/{path}"
    elif "/" in u and not u.startswith("http") and "." not in u.split("/")[0]:
        u = f"https://github.com/{u}"
    lowered = u.lower()
    prefix = "https://github.com/"
    if not lowered.startswith(prefix):
        return None
    tail = lowered.removeprefix(prefix).strip("/")
    if "/" not in tail:
        return None
    owner, repo = tail.split("/", 1)
    owner = owner.strip().lower()
    repo = repo.strip().lower().removesuffix(".git")
    if not owner or not repo:
        return None
    return f"https://github.com/{owner}/{repo}"


def _github_owner(repo_url: str) -> str | None:
    norm = normalize_github_https(repo_url)
    if not norm:
        return None
    rest = norm.removeprefix("https://github.com/")
    return rest.split("/", 1)[0].lower()


def is_restrictive_mode() -> bool:
    return bool(
        _parse_csv("INDEXER_ALLOWED_GITHUB_ORGS") or _parse_csv("INDEXER_ALLOWED_REPO_URLS")
    )


def assert_repo_url_allowed(repo_url: str) -> None:
    """
    Raise ValueError with a short message if repo_url is not allowed.

    Used by HTTP /index and batch_index before any clone.
    """
    if not is_repo_url_allowed(repo_url):
        raise ValueError(f"Repository URL not allowed by policy: {repo_url!r}")


def is_repo_url_allowed(repo_url: str) -> bool:
    """Return True if indexing this repo_url is permitted."""
    raw = repo_url.strip()
    if not raw:
        return False

    if raw.lower().startswith("file:"):
        if os.environ.get("INDEXER_ALLOW_FILE_REMOTES", "").strip().lower() not in (
            "1",
            "true",
            "yes",
        ):
            logger.warning("file:// repo rejected: INDEXER_ALLOW_FILE_REMOTES not enabled")
            return False
        parsed = urlparse(raw)
        path = parsed.path or ""
        if not path:
            return False
        prefixes = _parse_path_prefixes()
        ok = any(path.startswith(p) for p in prefixes)
        if not ok:
            logger.warning("file:// path %s not under allowed prefixes %s", path, prefixes)
        return ok

    norm = normalize_github_https(raw)
    if norm is None:
        logger.warning("Non-GitHub HTTPS URL rejected: %s", raw)
        return False

    if not is_restrictive_mode():
        return True

    orgs = {o.lower() for o in _parse_csv("INDEXER_ALLOWED_GITHUB_ORGS")}
    urls = set()
    for entry in _parse_csv("INDEXER_ALLOWED_REPO_URLS"):
        n = normalize_github_https(entry)
        if n:
            urls.add(n)

    org_ok = not orgs or (_github_owner(raw) in orgs)
    url_ok = not urls or (norm in urls)
    return org_ok and url_ok

File: services/git-watcher/src/main/kotlin/com/k8sqarag/gitwatcher/RepoAllowlist.kt
package com.k8sqarag.gitwatcher

/**
 * Matches [INDEXER_ALLOWED_GITHUB_ORGS] and [INDEXER_ALLOWED_REPO_URLS] semantics
 * documented with the Python indexer (`services/indexer/repo_allowlist.py`).
 *
 * GitHub webhooks should only deliver https://github.com/... clone URLs.
 * [file://] is always denied at the watcher (use HTTPS remotes from GitHub only).
 */
object RepoAllowlist {

    fun matches(cloneUrl: String): Boolean {
        val raw = cloneUrl.trim()
        if (raw.isEmpty()) return false
        if (raw.startsWith("file:", ignoreCase = true)) return false

        val norm = normalizeGithubHttps(raw) ?: return false

        val orgs = parseCsv(System.getenv("INDEXER_ALLOWED_GITHUB_ORGS")).map { it.lowercase() }.toSet()
        val urlEntries = parseCsv(System.getenv("INDEXER_ALLOWED_REPO_URLS"))
        val urls = urlEntries.mapNotNull { normalizeGithubHttps(it) }.toSet()
        val restrictive = orgs.isNotEmpty() || urls.isNotEmpty()
        if (!restrictive) return true

        val owner = githubOwner(norm) ?: return false
        val orgOk = orgs.isEmpty() || orgs.contains(owner)
        val urlOk = urls.isEmpty() || urls.contains(norm)
        return orgOk && urlOk
    }

    private fun parseCsv(value: String?): List<String> =
        value
            ?.split(",")
            ?.map { it.trim() }
            ?.filter { it.isNotEmpty() }
            ?: emptyList()

    fun normalizeGithubHttps(repoUrl: String): String? {
        var u = repoUrl.trim().trimEnd('/')
        if (u.endsWith(".git", ignoreCase = true)) u = u.dropLast(4)
        val lower = u.lowercase()
        if (lower.startsWith("git@github.com:")) {
            u = Regex("^git@github\\.com:", RegexOption.IGNORE_CASE).replace(u, "https://github.com/")
            u = u.trimEnd('/')
        } else if (!lower.startsWith("http") && u.contains("/") && !u.contains("://")) {
            u = "https://github.com/${u.trim('/')}"
        }
        if (!u.startsWith("https://github.com/", ignoreCase = true)) return null
        val tail = u.substring("https://github.com/".length).trim('/')
        val parts = tail.split('/').filter { it.isNotEmpty() }
        if (parts.size < 2) return null
        val owner = parts[0].lowercase()
        val repo = parts[1].lowercase().removeSuffix(".git")
        if (owner.isEmpty() || repo.isEmpty()) return null
        return "https://github.com/$owner/$repo"
    }

    private fun githubOwner(normalized: String): String? {
        val rest = normalized.removePrefix("https://github.com/").trim('/')
        return rest.split('/').firstOrNull()?.lowercase()
    }
}
